flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/utils/DataSetConversionUtil.java

```diff
@@ -0,0 +1,172 @@
+			.toDataSet(table, Row.class);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified TableSchema.
+	 *
+	 * @param sessionId the sessionId of {@link MLEnvironmentFactory}
+	 * @param data   the DataSet to convert.
+	 * @param schema the specified TableSchema.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, TableSchema schema) {
+		return toTable(sessionId, data, schema.getFieldNames(), schema.getFieldTypes());
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames and colTypes.
+	 *
+	 * @param sessionId sessionId the sessionId of {@link MLEnvironmentFactory}.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @param colTypes the specified colTypes. This variable is used only when the
+	 *                 DataSet is produced by a function and Flink cannot determine
+	 *                 automatically what the produced type is.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, String[] colNames, TypeInformation <?>[] colTypes) {
+		return toTable(MLEnvironmentFactory.get(sessionId), data, colNames, colTypes);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames.
+	 *
+	 * @param sessionId sessionId the sessionId of {@link MLEnvironmentFactory}.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(Long sessionId, DataSet <Row> data, String[] colNames) {
+		return toTable(MLEnvironmentFactory.get(sessionId), data, colNames);
+	}
+
+	/**
+	 * Convert the given DataSet into a Table with specified colNames and colTypes.
+	 *
+	 * @param session the MLEnvironment using to convert DataSet to Table.
+	 * @param data     the DataSet to convert.
+	 * @param colNames the specified colNames.
+	 * @param colTypes the specified colTypes. This variable is used only when the
+	 *                 DataSet is produced by a function and Flink cannot determine
+	 *                 automatically what the produced type is.
+	 * @return the converted Table.
+	 */
+	public static Table toTable(MLEnvironment session, DataSet <Row> data, String[] colNames, TypeInformation <?>[] colTypes) {
+		try {
+			// Try to add row type information for the dataset to be converted.
+			// In most case, this keeps us from the rolling back logic in the catch block,
+			// which adds an unnecessary map function just in order to add row type information.
+			if (data instanceof SingleInputUdfOperator) {
+				((SingleInputUdfOperator) data).returns(new RowTypeInfo(colTypes, colNames));
+			} else if (data instanceof TwoInputUdfOperator) {
+				((TwoInputUdfOperator) data).returns(new RowTypeInfo(colTypes, colNames));
+			}
+			return toTable(session, data, colNames);
+		} catch (ValidationException ex) {
+			if (null == colTypes) {
```

Comment:
shouldn't we always honor users' explicit configuration of `colTypes` instead of trying to implicitly cast it?

Reply:
This function is used in all operators, so if we use the explicit configuration of colTypes , the `mapfunction` will be found everywhere, which rise concern for performance.

Reply:
oops.. that's a good point! this will in fact create performance issue in runtime.

Reply:
so, how about keep the original implementation?

Reply:
Yes. I think for now we can merge it as is and improve it later - I will add some comments.